### DISCUSSION TITLE
importing a package before generating

### DISCUSSION THREAD

**Discussion by pmonty** [babel/babel#14237] [Q&A]
Hi all, so I have got an app that uses @babel/standalone with a combination of escodegen and acorn to render some JSX that is retrieved at runtime to compile and display a small reactjs form. 

It is working great but I did want to use some react packages like react-hook-form for example. Issue is unless its vanilla react it doesn't work. The moment I try import a library i get errors. I haven't attempted in a while cause I was stuck on the issue for over a week previously but will try replicate and provide them.

Vanilla JS works though... for example
```
function ContactForm() {
    const [state, setState] = React.useState(injectstatehere);
    const [submitted, setSubmitted] = React.useState(false);
    const [documents, setDocuments] = React.useState([]);

    const handleChange = (e) => {
      const { name, value } = e.target;
      setState((prevState) => ({
        ...prevState,
        [name]: value,
      }));
    };
    
    const generateFieldsSubmission = () => {
      // ...
    };

    const handleSubmit = (e) => {
       // ...
    };
  
    return (<>
        <p>test</p>
      </>
    );
};
  
<ContactForm />
```

I guess what I am looking for is the ability to load modules that are installed in the parent react app. I am also happy to have to bundle up packages and host them myself so that they can be referenced in a <script /> tag myself and then in the html page include that. But have a feeling that won't work since it will be 2 levels away...
level 1 HTML
level 2 React App
level 3 babel/standalone script

UPDATE:
Found this cool [sandbox ](https://codesandbox.io/s/dikoh?file=/index.html:516-556) using skypack.dev... might try this! looks promising

**Comment by pmonty** [comment]
No luck wit the skypack.dev setup that I attached in initial post. Has anyone got something working with a setup like that?

The reason it won't work with my setup is because I don't have the opportunity to include it in a script tag at buildtime. It's dynamic and the script is retrieved by the parent react app so I am using an approach like this https://codesandbox.io/s/jgm04?file=/src/editor.js:1035-1054

**Comment by nicolo-ribaudo** [comment]
`@babel/standalone` does not support running native modules in the browser: you either have to compile them to something like AMD/RequrieJS, or  you have to pre-bundle your files on the server.

**Reply by pmonty** [reply]
Hi @nicolo-ribaudo is that skypack example I shared not the way to include other packages? 
https://codesandbox.io/s/dikoh?file=/index.html:262-1106

I'm not sure if it works only because it's in a script tag and not being done at runtime. If that is the reason then I might be stuck and need to rethink everything.

Edit: Also if this is not possible at the moment, could it be eventually?

**Reply by pmonty** [reply]
Actually looks like https://github.com/babel/babel/discussions/12059 might have accomplished something here using a serviceworker file. Or am I not looking at this correctly?

**Reply by nicolo-ribaudo** [reply]
That skypack example works because you are using a modern browser that natively supports all the syntax you need: Babel is not transpiring the imported modules.

Something like the service worker example you found might work (we may even implement something similar in the future), but for now Babel does not support compiling on-the-fly files loaded with ESM imports.

**Reply by pmonty** [reply]
Hi @nicolo-ribaudo just wondering if there is an active discussion or thread somewhere that talks about adding ESM import support? I am still working at using a service-worker to do this but wondering how the official support discussion is going or if there is a public thread somewhere to follow?
